pkg/agent/core/state.go

```diff
@@ -791,12 +850,59 @@ func (s *state) desiredResourcesFromMetricsOrRequestedUpscaling(now time.Time) (
 			return nil
 		}
 	}
+	s.updateTargetRevision(now, result, s.VM.Using())
 
-	s.info("Calculated desired resources", zap.Object("current", s.VM.Using()), zap.Object("target", result))
+	// TODO: we are both saving the result into LastDesiredResources and returning it. This is
+	// redundant, and we should remove one of the two.
+	s.LastDesiredResources = &result
+
+	s.info("Calculated desired resources",
+		zap.Object("current", s.VM.Using()),
+		zap.Object("target", result),
+		zap.Object("targetRevision", &s.TargetRevision))
 
 	return result, calculateWaitTime
 }
 
+func (s *state) updateTargetRevision(now time.Time, desired api.Resources, current api.Resources) {
+	if s.LastDesiredResources == nil {
+		s.LastDesiredResources = &current
+	}
+
+	if *s.LastDesiredResources == desired {
+		// Nothing changed, so no need to update the target revision
+		return
+	}
+
+	var flags vmv1.Flag
+
+	if desired.HasFieldGreaterThan(*s.LastDesiredResources) {
+		flags.Set(revsource.Upscale)
+	}
+	if desired.HasFieldLessThan(*s.LastDesiredResources) {
+		flags.Set(revsource.Downscale)
+	}
+
+	s.TargetRevision = s.Config.RevisionSource.Next(now, flags)
+}
+
+func (s *state) updateNeonVMCurrentRevision(currentRevision vmv1.RevisionWithTime) {
+	revsource.Propagate(currentRevision.UpdatedAt.Time,
+		s.NeonVM.TargetRevision,
+		&s.NeonVM.CurrentRevision,
+		s.Config.ObservabilityCallbacks.NeonVMLatency,
+	)
+	err := s.Config.RevisionSource.Observe(currentRevision.UpdatedAt.Time, currentRevision.Revision)
+
+	// We also zero out LastDesiredResources, because we are now starting from
+	// a new current resources.
+	s.LastDesiredResources = nil
```

Comment:
Should we only do this when `LastDesiredResources == current neonvm resources` ?

Reply:
discussed -- the current behavior is actually required.